internal/controller/postgrescluster/instance.go

```diff
@@ -1341,3 +1384,361 @@ func (r *Reconciler) reconcileInstanceCertificates(
+	var primary *Instance
+	for i, instance := range observed.forCluster {
+		if !clusterRunning {
+			clusterRunning, _ = instance.IsRunning(naming.ContainerDatabase)
+		}
+		if instance.Runner != nil {
+			runners = append(runners, instance.Runner)
+		}
+		if isPrimary, _ := instance.IsPrimary(); isPrimary {
+			primary = observed.forCluster[i]
+		}
+	}
+
+	// Set the proper startup instance for the upgrade. The primary is required and
+	// all replicas will be scaled down.
+	if cluster.Status.StartupInstanceSet == "" || cluster.Status.StartupInstance == "" {
+		if primary != nil {
+			cluster.Status.StartupInstance = primary.Name
+			cluster.Status.StartupInstanceSet = primary.Spec.Name
+		} else if len(runners) > 0 {
+			cluster.Status.StartupInstance = runners[0].GetName()
+			cluster.Status.StartupInstanceSet =
+				runners[0].GetLabels()[naming.LabelInstanceSet]
+		} else if len(cluster.Spec.InstanceSets) > 0 {
+			// Generate a hash that will be used make sure that the startup
+			// instance is named consistently
+			cluster.Status.StartupInstance = naming.GenerateStartupInstance(cluster,
+				&cluster.Spec.InstanceSets[0]).Name
+			cluster.Status.StartupInstanceSet = cluster.Spec.InstanceSets[0].Name
+		} else {
+			return errors.New("unable to determine startup instance for major upgrade")
+		}
+	}
+
+	// remove any existing upgrade Jobs
+	if upgradeJob != nil {
+		setPreparingClusterCondition("removing existing upgrade job")
+		if err := r.Client.Delete(ctx, upgradeJob,
+			client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+			return errors.WithStack(err)
+		}
+		return nil
+	}
+
+	if clusterRunning {
+		setPreparingClusterCondition("removing runners")
+		for _, runner := range runners {
+			err := r.Client.Delete(ctx, runner,
+				client.PropagationPolicy(metav1.DeletePropagationForeground))
+			if client.IgnoreNotFound(err) != nil {
+				return errors.WithStack(err)
+			}
+		}
+		return nil
+	}
+
+	// if everything is gone, proceed with re-bootstrapping the cluster
+	if len(currentEndpoints) == 0 {
+		if len(cluster.Status.Conditions) > 0 {
+			// TODO: remove guard with move to controller-runtime 0.9.0 https://issue.k8s.io/99714
+			meta.RemoveStatusCondition(&cluster.Status.Conditions, ConditionPostgresDataInitialized)
+		}
+		meta.SetStatusCondition(&cluster.Status.Conditions, metav1.Condition{
+			ObservedGeneration: cluster.GetGeneration(),
+			Type:               ConditionPGUpgradeProgressing,
+			Status:             metav1.ConditionTrue,
+			Reason:             ReasonReadyForUpgrade,
+			Message:            "Upgrading cluster postgres major version",
+		})
+		// the cluster is no longer bootstrapped
+		cluster.Status.Patroni.SystemIdentifier = ""
+		// the upgrade will change the contents of the database, so the pgbouncer and exporter hashes
+		// are no longer valid
+		cluster.Status.Proxy.PGBouncer.PostgreSQLRevision = ""
+		cluster.Status.Monitoring.ExporterConfiguration = ""
```

Comment:
After discussing with the team, for now we'll plan to leave the status as is. Depending on the particulars of the custom sql configured, it may not be desirable to automatically run again.